drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -0,0 +1,341 @@
+	}
+
+	/* disable adc interrupt */
+	irq_disable(DT_INST_IRQN(0));
+}
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_it8xxx2_start_read(const struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	LOG_DBG("Configure resolution=%d", sequence->resolution);
+
+	if (!sequence->channels ||
+		sequence->channels & ~BIT_MASK(CHIP_ADC_COUNT)) {
+		LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	err = check_buffer_size(sequence, 1);
+	if (err) {
+		return err;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	/* enable adc interrupt */
+	irq_enable(DT_INST_IRQN(0));
+
+	/* bit 0 : adc module enable */
+	IT83XX_ADC_ADCCFG |= 0x01;
+}
+
+static int adc_it8xxx2_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	err = adc_it8xxx2_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_data_valid(enum chip_adc_channel adc_ch)
+{
+	return IT83XX_ADC_ADCDVSTS & BIT(adc_ch);
+}
+
+static void adc_it8xxx2_get_sample(const struct device *dev)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	uint8_t channels = data->ctx.sequence.channels;
+	uint8_t idx = 0;
+	uint8_t valid = 0;
+	uint8_t bit;
+
+	bit = find_lsb_set(channels);
+	while (bit != 0) {
+		idx = bit - 1;
+		if (adc_data_valid(idx)) {
+
+			*data->buffer = ((*adc_ctrl_regs[idx].adc_datm << 8) +
+						*adc_ctrl_regs[idx].adc_datl) *
+						data->ctx.sequence.resolution;
+
+			data->buffer++;
+
+			/* W/C data valid flag */
+			IT83XX_ADC_ADCDVSTS = BIT(idx);
+			valid = 1;
+		}
+
+		if (!valid) {
+			printk("ADC failed to read!!! (regs=%x, ch=%d)",
+				IT83XX_ADC_ADCDVSTS, idx);
+		}
+
+		channels &= ~BIT(idx);
+		bit = find_lsb_set(channels);
+		/* disable adc interrupt */
+		adc_disable_channel(dev, idx);
```

Comment:
ADC interrupt for particular channels is disabled after reading one sample. But if I understand it correctly, the interrupt is enabled for a given channel in [`adc_it8xxx2_channel_setup`](https://github.com/zephyrproject-rtos/zephyr/blob/f00dd9a886d82a011f553a4c0d971c5ec0aba352/drivers/adc/adc_ite_it8xxx2.c#L88-L106). If so, it won't work if `sequence->options->extra_samplings` is non-zero (see e.g. `test_adc_repeated_samplings()` in `tests/drivers/adc/adc_api`) or if `adc_read()` is called again without `adc_channel_setup()` before it. Or am I missing something?

Reply:
In the previous version, to enable interrupt has been changed from adc_it8xxx2_channel_setup() to adc_context_start_sampling(), it will pass all test of tests/drivers/adc/adc_api including test_ade_repeated_samplings().

But in adc_disable_channel(), the register of clearing ADC control needs to be deleted because it will be as you said adc_read() is called again without adc_channel_setup().